extensions/wikia/UserLogin/tests/UserLoginTest.php

```diff
@@ -1,782 +1,869 @@
-				],
-				[
-					'query' => [],
-					'wfArrayToCGI' => '',
-					'wfCgiToArrayResult' => [],
-					'expected' => '',
-					'message' => 'Query without any parameters',
-				],
-				[
-					'query' => [
-						'login' => self::TEST_USERNAME,
-						'returnto' => 'Special:UserLogin',
-						'editToken' => '123456789',
-					],
-					'wfArrayToCGI' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
-					'wfCgiToArrayResult' => [],
-					'expected' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
-					'message' => 'Query without returntoquery parameter',
+		$result = $getReturnToQueryFromQueryMethod->invoke( new UserLoginSpecialController(), $query );
+		$this->assertEquals( $expected, $result, $message );
+	}
+
+	public function getReturnToQueryFromQueryDataProvider() {
+		return [
+			[
+				'query' => '',
+				'wfArrayToCGI' => '',
+				'wfCgiToArrayResult' => [],
+				'expected' => '',
+				'message' => 'Query is not an array',
+			],
+			[
+				'query' => [],
+				'wfArrayToCGI' => '',
+				'wfCgiToArrayResult' => [],
+				'expected' => '',
+				'message' => 'Query without any parameters',
+			],
+			[
+				'query' => [
+					'login' => self::TEST_USERNAME,
+					'returnto' => 'Special:UserLogin',
+					'editToken' => '123456789',
 				],
-				[
-					'query' => [
-						'login' => self::TEST_USERNAME,
-						'returnto' => 'Special:UserLogin',
-						// notice the returnto&login in previous returntoquery were different
-						// and are supposed to be overwritten
-						'returntoquery' => 'returnto=Main_page%3Alogin=Test%3AeditToken=123456789',
-						'editToken' => '123456789',
-					],
-					'wfArrayToCGI' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
-					'wfCgiToArrayResult' => [],
-					'expected' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
-					'message' => 'Query with returntoquery',
+				'wfArrayToCGI' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
+				'wfCgiToArrayResult' => [],
+				'expected' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
+				'message' => 'Query without returntoquery parameter',
+			],
+			[
+				'query' => [
+					'login' => self::TEST_USERNAME,
+					'returnto' => 'Special:UserLogin',
+					// notice the returnto&login in previous returntoquery were different
+					// and are supposed to be overwritten
+					'returntoquery' => 'returnto=Main_page%3Alogin=Test%3AeditToken=123456789',
+					'editToken' => '123456789',
 				],
-			];
-		}
+				'wfArrayToCGI' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
+				'wfCgiToArrayResult' => [],
+				'expected' => 'login=' . self::TEST_USERNAME . '&returnto=Special%3AUserLogin&editToken=123456789',
+				'message' => 'Query with returntoquery',
+			],
+		];
+	}
+
```

Comment:
Since `getReturnToUrl()` is removed, `testGetRedirectURL` and `getRedirectURLProvider` would be better names.